### PR TITLE
Added type property to node

### PR DESCRIPTION
**Pull request by Haaroon**
### What changes were proposed in this pull request?

Nodes can now have explicit types! 

### Why are the changes needed?

More flexibility

### Does this PR introduce any user-facing change? If yes is this documented?

Yes all docs

### How was this patch tested?

With tests

### CODE REVIEW

**Comment by Haaroon** [comment]
[#1481](https://github.com/Pometry/Raphtory/issues/1481)

File: raphtory/src/search/mod.rs [hunk 1]
```diff
@@ -804,7 +808,7 @@ impl<G: StaticGraphViewOps + InternalAdditionOps> InternalAdditionOps for Indexe
             }
         }
         // add the node id to the document
-        self.graph.internal_add_node(t, v, props)?;
+        self.graph.internal_add_node(t, v, props, node_type_id)?;
         // get the field from the index
         let node_id = self.node_index.schema().get_field(fields::VERTEX_ID)?;
         let node_id_rev = self.node_index.schema().get_field(fields::VERTEX_ID_REV)?;
```
Comment (ljeub-pometry) [inline raphtory/src/search/mod.rs:814]: should make the node type searchable

**Review by ljeub-pometry — CHANGES_REQUESTED** [review]
Some minor changes:

- make node type return `Option<ArcStr>` to avoid cloning the actual string and remove the magic `"_default"` node type. 
- make changing the node type an error for the add_node method and add an explicit update_type method on the mutable node view 
- add a test for pandas loading without node type column
- improve `get_node_types` on graph and expose the functionality in the public graph view
- consider improving the import by using the internal methods and only doing the string resolution once (optional, make an issue for later otherwise)
- we should make node type searchable (optional, make an issue for later otherwise)

File: raphtory/src/core/entities/properties/props.rs [hunk 2]
```diff
@@ -187,6 +214,14 @@ impl Meta {
             .collect()
     }
 
+    pub fn get_all_node_types(&self) -> Vec<usize> {
```
[the remark this answers is not in the document]
Comment (ljeub-pometry) [reply raphtory/src/core/entities/properties/props.rs:217]: probably better, make this return the type names instead by calling `get_keys()`

File: raphtory/src/core/entities/properties/props.rs [hunk 3]
```diff
@@ -202,8 +202,12 @@ impl Meta {
...
     }
 
-    pub fn get_node_type_name_by_id(&self, id: usize) -> ArcStr {
-        self.meta_node_type.get_name(id)
+    pub fn get_node_type_name_by_id(&self, id: usize) -> Option<ArcStr> {
+        if self.meta_node_type.has_name(id) && self.meta_node_type.get_name(id) != "_default" {
+            Some(self.meta_node_type.get_name(id))
+        } else {
+            None
+        }
```
Comment (ljeub-pometry) [inline raphtory/src/core/entities/properties/props.rs:210]: The assumption is that if you have an internal id, the name entry exists. 

this should be 
```
if id == 0 {
    None
} else {
    Some(self.meta_node_type.get_name(id))
}
```

File: raphtory/src/db/api/mutation/addition_ops.rs [hunk 4]
```diff
@@ -114,7 +114,10 @@ impl<G: InternalAdditionOps + StaticGraphViewOps> AdditionOps for G {
         )?;
         let ti = TimeIndexEntry::from_input(self, t)?;
         let v_id = self.resolve_node(v.id(), v.id_str());
-        let type_id = self.resolve_node_type(v_id, node_type);
+        let type_id = self.resolve_node_type(v_id, node_type).expect(&format!(
+            "Error when running resolve_node_type {:?}",
+            node_type
+        ));
```
Comment (ljeub-pometry) [inline raphtory/src/db/api/mutation/addition_ops.rs:120]: this should not be an expect, just `?`

File: raphtory/src/db/api/mutation/import_ops.rs [hunk 5]
```diff
@@ -126,16 +127,39 @@ impl<
             }
         }
 
+        let node_internal =
+            self.resolve_node(node.id(), node.graph.core_node(node.node).name.as_str());
+        let node_internal_type_id = self
+            .resolve_node_type(node_internal, node.node_type().as_str())
+            .unwrap_or(0usize);
```
Comment (ljeub-pometry) [inline raphtory/src/db/api/mutation/import_ops.rs:134]: don't unwrap, just error if this fails

File: raphtory/src/db/api/view/graph.rs [hunk 6]
```diff
@@ -152,12 +152,19 @@ impl<'graph, G: BoxableGraphView<'graph> + Sized + Clone + 'graph> GraphViewOps<
         }
 
         for v in self.nodes().iter() {
+            let v_type_string = v.node_type(); //.map(|v| v.to_string()); //stop it being dropped
+            let v_type_str = v_type_string.as_str();
```
Comment (ljeub-pometry) [inline raphtory/src/db/api/view/graph.rs:156]: If we make `node_type()` return `Option<&ArcStr>` we wouldn't have to do this dance.

File: raphtory/src/db/api/view/graph.rs [hunk 7]
```diff
@@ -327,8 +334,14 @@ mod test_materialize {
         let g = Graph::new();
         let node_a = g.add_node(0, "A", NO_PROPS, None).unwrap();
         let node_b = g.add_node(1, "B", NO_PROPS, Some(&"H")).unwrap();
-        assert_eq!(node_a.node_type(), "_default");
-        assert_eq!(node_b.node_type(), "H");
+        let node_a_ref = node_a.node_type().map(|v| v.to_string()); //stop it being dropped
+        let node_a_type_str = node_a_ref.as_deref();
```
Comment (ljeub-pometry) [inline raphtory/src/db/api/view/graph.rs:338]: `.as_str()`

File: raphtory/src/db/api/view/graph.rs [hunk 8]
```diff
@@ -327,8 +334,14 @@ mod test_materialize {
...
+
+        assert_eq!(node_a_type_str, None);
+        assert_eq!(
+            node_b.node_type().map(|v| v.to_string()).as_deref(),
+            Some("H")
+        );
```
Comment (ljeub-pometry) [inline raphtory/src/db/api/view/graph.rs:344]: `as_str()`

File: raphtory/src/db/api/view/internal/core_ops.rs [hunk 9]
```diff
@@ -47,7 +47,7 @@ pub trait CoreGraphOps {
     fn node_name(&self, v: VID) -> String;
 
     /// Returns the type of node
-    fn node_type(&self, v: VID) -> String;
+    fn node_type(&self, v: VID) -> Option<ArcStr>;
```
Comment (ljeub-pometry) [inline raphtory/src/db/api/view/internal/core_ops.rs:50]: making this `Option<&ArcStr>` will make our life easier

File: raphtory/src/db/api/view/internal/core_ops.rs [hunk 10]
```diff
@@ -271,7 +271,7 @@ impl<G: DelegateCoreOps + ?Sized> CoreGraphOps for G {
     }
 
     #[inline]
-    fn node_type(&self, v: VID) -> String {
+    fn node_type(&self, v: VID) -> Option<ArcStr> {
```
Comment (ljeub-pometry) [inline raphtory/src/db/api/view/internal/core_ops.rs:274]: making this Option<&ArcStr> will make our life easier

File: raphtory/src/db/api/view/node.rs [hunk 11]
```diff
@@ -74,7 +74,7 @@ pub trait NodeViewOps<'graph>: Clone + TimeOps<'graph> + LayerOps<'graph> {
     fn name(&self) -> Self::ValueType<String>;
 
     /// Returns the type of node
-    fn node_type(&self) -> Self::ValueType<String>;
+    fn node_type(&self) -> Self::ValueType<Option<ArcStr>>;
```
Comment (ljeub-pometry) [inline raphtory/src/db/api/view/node.rs:77]: making this Option<&ArcStr> will make our life easier

File: raphtory/src/db/api/view/node.rs [hunk 12]
```diff
@@ -180,7 +180,7 @@ impl<'graph, V: BaseNodeViewOps<'graph> + 'graph> NodeViewOps<'graph> for V {
         self.map(|g, v| g.node_name(v))
     }
     #[inline]
-    fn node_type(&self) -> Self::ValueType<String> {
+    fn node_type(&self) -> Self::ValueType<Option<ArcStr>> {
```
Comment (ljeub-pometry) [inline raphtory/src/db/api/view/node.rs:183]: making this Option<&ArcStr> will make our life easier

File: raphtory/src/db/internal/core_ops.rs [hunk 13]
```diff
@@ -51,7 +51,7 @@ impl<const N: usize> CoreGraphOps for InnerTemporalGraph<N> {
     }
 
     #[inline]
-    fn node_type(&self, v: VID) -> String {
+    fn node_type(&self, v: VID) -> Option<ArcStr> {
```
Comment (ljeub-pometry) [inline raphtory/src/db/internal/core_ops.rs:54]: making this Option<&ArcStr> will make our life easier

**Review by ljeub-pometry — CHANGES_REQUESTED** [review]
`resolve_node_type` still has some logic issues and the related error handling needs to be fixed. We should also add some tests for this (i.e., make sure that overwriting the type errors and doesn't silently fail or change the type and also that a failed update does not actually add the type to the graph).

File: raphtory/src/core/entities/graph/tgraph.rs [hunk 14]
```diff
@@ -362,14 +358,33 @@ impl<const N: usize> TemporalGraph<N> {
...
+                match node.node_type {
+                    0 => {
+                        node.update_node_type(node_type_id);
+                        Ok(node_type_id)
+                    }
+                    _ => {
+                        // Returns the original node type to prevent type being changed
+                        Ok(node.node_type)
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (Haaroon) [reply raphtory/src/core/entities/graph/tgraph.rs:387]: add_node_no_props
 calls update_node_type which is why i have the checks in update_node_type,

**Comment by Haaroon** [comment]
just to note, we cant change the ArcStr to &ArcStr due to the lifetime issues, this is why it was not done before

**Pull request merged**